src/input_common/drivers/touch_screen.cpp

```diff
@@ -14,38 +14,92 @@ constexpr PadIdentifier identifier = {
+    ReleaseAllTouch();
 }
 
-void TouchScreen::TouchMoved(float x, float y, std::size_t finger) {
-    if (finger >= 16) {
+void TouchScreen::TouchMoved(float x, float y, std::size_t finger_id) {
+    const auto index = GetIndexFromFingerId(finger_id);
+    if (index == MAX_FINGER_COUNT) {
+        // Touch doesn't exist handle it as a new one
+        TouchPressed(x, y, finger_id);
         return;
     }
-    TouchPressed(x, y, finger);
+    fingers[index].is_active = true;
+    SetButton(identifier, static_cast<int>(index), true);
+    SetAxis(identifier, static_cast<int>(index * 2), x);
+    SetAxis(identifier, static_cast<int>(index * 2 + 1), y);
 }
 
-void TouchScreen::TouchPressed(float x, float y, std::size_t finger) {
-    if (finger >= 16) {
+void TouchScreen::TouchPressed(float x, float y, std::size_t finger_id) {
+    auto index = GetIndexFromFingerId(finger_id);
+    if (index != MAX_FINGER_COUNT) {
+        // Touch already exist. Just update the data
+        TouchMoved(x, y, finger_id);
         return;
     }
-    SetButton(identifier, static_cast<int>(finger), true);
-    SetAxis(identifier, static_cast<int>(finger * 2), x);
-    SetAxis(identifier, static_cast<int>(finger * 2 + 1), y);
+    index = GetNextFreeIndex();
+    if (index == MAX_FINGER_COUNT) {
+        // No free entries. Ignore input
+        return;
+    }
+    fingers[index].is_enabled = true;
+    fingers[index].finger_id = finger_id;
+    TouchMoved(x, y, finger_id);
 }
 
-void TouchScreen::TouchReleased(std::size_t finger) {
-    if (finger >= 16) {
+void TouchScreen::TouchReleased(std::size_t finger_id) {
+    const auto index = GetIndexFromFingerId(finger_id);
+    if (index == MAX_FINGER_COUNT) {
         return;
     }
-    SetButton(identifier, static_cast<int>(finger), false);
-    SetAxis(identifier, static_cast<int>(finger * 2), 0.0f);
-    SetAxis(identifier, static_cast<int>(finger * 2 + 1), 0.0f);
+    fingers[index].is_enabled = false;
+    SetButton(identifier, static_cast<int>(index), false);
+    SetAxis(identifier, static_cast<int>(index * 2), 0.0f);
+    SetAxis(identifier, static_cast<int>(index * 2 + 1), 0.0f);
+}
+
+std::size_t TouchScreen::GetIndexFromFingerId(std::size_t finger_id) const {
+    for (std::size_t index = 0; index < MAX_FINGER_COUNT; ++index) {
+        const auto& finger = fingers[index];
+        if (!finger.is_enabled) {
+            continue;
+        }
+        if (finger.finger_id == finger_id) {
+            return index;
+        }
+    }
+    return MAX_FINGER_COUNT;
+}
+
+std::size_t TouchScreen::GetNextFreeIndex() const {
+    for (std::size_t index = 0; index < MAX_FINGER_COUNT; ++index) {
+        if (!fingers[index].is_enabled) {
+            return index;
+        }
+    }
+    return MAX_FINGER_COUNT;
+}
+
+void TouchScreen::ClearActiveFlag() {
+    for (auto& finger : fingers) {
+        finger.is_active = false;
+    }
+}
+
+void TouchScreen::ReleaseInactiveTouch() {
+    for (auto& finger : fingers) {
```

Comment:
```suggestion
    for (const auto& finger : fingers) {
```

since the reference isn't modified